xdsmatcher/pkg/matcher/matcher.go

```diff
@@ -0,0 +1,523 @@
+package matcher
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	pbmatcher "github.com/cncf/xds/go/xds/type/matcher/v3"
+	pblegacymatcher "github.com/envoyproxy/go-control-plane/envoy/config/common/matcher/v3"
+	"github.com/envoyproxy/go-control-plane/xdsmatcher/pkg/matcher/registry"
+	"github.com/envoyproxy/go-control-plane/xdsmatcher/pkg/matcher/types"
+	"google.golang.org/protobuf/proto"
+)
+
+type MatcherTree struct {
+	onNoMatch *types.OnMatch
+	matchRoot types.Matcher
+
+	logger *tracingLogger
+}
+
+func (m MatcherTree) Match(data types.MatchingData) (types.Result, error) {
+	result, err := m.matchRoot.Match(data)
+	if err != nil {
+		return types.Result{}, err
+	}
+	if result.MatchResult != nil && result.NeedMoreData {
+		return types.Result{}, errors.New("invalid result, invariant violation")
+	}
+
+	// We arrived at a result.
+	if result.MatchResult != nil {
+		// The result is a recursive matcher, so recurse.
+		if result.MatchResult.Matcher != nil {
+			m.logger.log("recursively matching")
+			m.logger.push()
+			defer m.logger.pop()
+			return result.MatchResult.Matcher.Match(data)
+		}
+
+		m.logger.log("arrived at result %+v, %+v", result, result.MatchResult)
+
+		// Otherwise pass through the result.
+		return result, nil
+	}
+
+	// We failed to arrive at a result due to the data not being available,
+	// report this back up.
+	if result.NeedMoreData {
+		m.logger.log("matcher tree requires more data")
+		return result, nil
```

Comment:
as i understand it a result with "needs more data" is valid, might just need a small comment adjustment

Reply:
Yeah this is not an error, it's the complex data type understanding that this data may change in the future, so it's unable to conclusively say that the data doesn't match.

An example here is how this is used for matching HTTP streams in Envoy: if you run into a matcher that tries to match against trailers before trailers are seen, we respond "NeedMoreData" so that the caller knows to call this again when trailers are made available.